designs/default-zero-value.md

```diff
@@ -0,0 +1,857 @@
+follows:
+
+- The `@default` trait can never be removed from a member.
+- The `@default` trait can only be added to a member if the member was
+  previously marked as `@required`.
+- The `@required` trait cannot be added to a member. This would transition the
+  member from nullable to non-nullable in generated code.
+- The `@required` trait can only be removed under the following conditions:
+  - It is replaced with the `@default` trait
+  - The containing structure is marked with the `@input` trait, meaning it is
+    only used as the input of a single operation. This affordance is only given
+    to top-level members of structures marked with the `@input` trait, and it
+    does not apply to nested input members.
+
+For example, if on _day-1_ the following structure is released:
+
+```
+structure Message {
+    @required
+    title: String
+
+    message: String
+}
+```
+
+On _day-2_, the service team realizes Message does not require a `title`, so
+they add the `@default` trait and remove the `@required` trait. The member will
+remain non-nullable because clients, servers, and any other deserializer will
+now provide a default value for the `title` member if a value is not provided by
+the end user.
+
+```
+structure Message {
+    @default
+    title: String
+
+    message: String
+}
+```
+
+Backward compatibility guarantees of the `@default` and `@required` traits
+SHOULD be enforced by tooling. For example, smithy-diff in Smithy's reference
+implementation will be updated to be aware of these additional backward
+compatibility rules.
+
+
+#### Special casing for @input structures
+
+The `@input` trait special-cases a structure as the input of a single
+operation. `@input` structures cannot be referenced in any other place in the
+model. Structures marked with the `@input` trait have more relaxed backward
+compatibility guarantees. It is backward compatible to remove the `@required`
+trait from top-level members of structures marked with the `@input` trait, and
+the `@required` trait does not need to be replaced with the `@default` trait.
+This gives service teams the ability to remove the `@required` trait from
+top-level input members without risking breaking previously generated clients.
+
+
+### `@default` trait syntactic sugar
+
+Smithy IDL 2.0 introduces syntactic sugar for attaching the `@default` trait to
+a member to make the importance of the `@default` trait more apparent, and to
+make it easier to find all structure members that are initialized to a default
+value. A member type followed by `= default` will be syntactic sugar for
+applying the `@default` trait to the member.
+
+The following structure:
+
+```
+structure Message {
+    title: String = default
+}
+```
+
+Is exactly equivalent to the following structure:
+
+```
+structure Message {
+    @default
+    title: String
+}
+```
+
+This pairs with the `@required` trait syntactic sugar that is also being added
```

Comment:
That looks good to me. It's nice because it's on the same line as the member, and these important traits have a special place to look for them.

Alternatively, I'm also open to just removing the syntactic sugar of both `!` and `= default` -- I worry a little that adding too much sugar will end up making it harder to use.

Reply:
Let's just yank this out. We can add sugar later if we feel we really need it.